crates/adapters/src/transport/pubsub/input.rs

```diff
@@ -0,0 +1,246 @@
+        let client_config = pubsub_config(config).await?;
+
+        let client = Client::new(client_config)
+            .await
+            .map_err(|e| anyhow!("error connecting to the Pub/Sub service: {e}"))?;
+
+        debug!(
+            "Pub/Sub input endpoint: connecting to subscription {}",
+            &config.subscription
+        );
+
+        let subscription = client.subscription(&config.subscription);
+        if !subscription.exists(None).await? {
+            bail!("subscription '{}' does not exist", config.subscription);
+        }
+
+        if let Some(snapshot) = &config.snapshot {
+            subscription
+                .seek(SeekTo::Snapshot(snapshot.to_string()), None)
+                .await
+                .map_err(|e| anyhow!("error retrieving Pub/Sub snapshot '{snapshot}': {e}"))?;
+        } else if let Some(timestamp) = &config.timestamp {
+            let ts = DateTime::parse_from_rfc3339(timestamp)
+                .map_err(|e| anyhow!("not a valid ISO 8601 date-time string '{timestamp}' {e}"))?;
+            subscription
+                .seek(SeekTo::Timestamp(SystemTime::from(ts)), None)
+                .await
+                .map_err(|e| anyhow!("error seeking to timestamp '{timestamp}': {e}"))?;
+        }
+
+        Ok(subscription)
+    }
+
+    async fn worker_task(
+        subscription: Subscription,
+        consumer: Box<dyn InputConsumer>,
+        mut state_receiver: Receiver<PipelineState>,
+    ) -> Result<(), AnyError> {
+        let mut state = PipelineState::Paused;
+        let mut cancel: Option<(CancellationToken, JoinHandle<()>)> = None;
+
+        loop {
+            // Wait for state change.
+            if let Ok(s) = state_receiver.wait_for(|s| s != &state).await {
+                state = *s;
+            } else {
+                warn!("Pub/Sub input endpoint: connector terminated unexpectedly, exiting worker task");
+                return Ok(());
+            };
+
+            match state {
+                PipelineState::Paused => {
+                    if let Some((token, handle)) = cancel.take() {
+                        token.cancel();
+                        let _ = handle.await;
+                    }
+                }
+                PipelineState::Running => {
+                    // TODO: Do we need to tune the config (the parameter to subscribe)?
+                    let mut stream = subscription
+                        .subscribe(None)
+                        .await
+                        .map_err(|e| anyhow!("error subscribing to messages: {e}"))?;
+
+                    let mut consumer_clone = consumer.clone();
+                    let token = stream.cancellable();
+                    let handle = tokio::spawn(async move {
+                        // None if the stream is cancelled
+                        while let Some(message) = stream.next().await {
+                            // println!(
```

Comment:
this could be an info logging statement?
we should be more consistent about this, it will make debugging in production easier later